routing/router.go

```diff
@@ -1101,3 +1104,54 @@ func (r *ChannelRouter) AddProof(chanID lnwire.ShortChannelID,
 	info.AuthProof = proof
 	return r.cfg.Graph.UpdateChannelEdge(info)
 }
+
+// getGraphDiameter is a function that determines the maximum distance between
+// two nodes in the graph, where the distance is calculated as the minimum number
+// of hops from src to dest
+func (r *ChannelRouter) getGraphDiameter() (uint8,error) {
+	// local variable to represent the graph diameter
+	var diam uint8
+	// If Cache is non-empty, then return Cached value
+	if r.diamCache != 0 {
+		// Acquire Mutex to read Cached value
+		r.diamCacheMtx.Lock()
+		diam = r.diamCache
+		r.diamCacheMtx.Unlock()
+		return diam
+	}
+	// x,y used to pass to findPath
+	x := make(map[vertex]struct{})
+	y := make(map[uint64]struct{})
+
+	if err := r.cfg.Graph.ForEachNode(nil, function (tx *bolt.Tx, src *channeldb.LightningNode) error {
```

Comment:
It seems like this algorithm is (in pseudo code):
```
for v1 in V:
    for v2 in V:
        diamPath = max(diamPath, len(findPath(v1, v2)))
```
Since `findPath(v1, v2)` runs a modified dijkstra's algorithm, it'll run in worst case `O(|V| log |V| + |E|)`.
Combined with the two outer loops over the vertices, our algorithm would run in worst case `O(|V|^3 log |V| + |V|^2 |E|)`.

However, since our graph is undirected and we need the longest shortest path from each node, we can run BFS from every node and record the distances, which we then max to find the longest, shortest path.

In pseudo code:
```
// BFS that records the shortest path from the start node to every other node
def BFS(start, shortestPathLen):
    shortestPathLen[start] = 0
    q = Queue()
    q.enqueue(start)
    while !q.empty():
        v1 = q.dequeue()
        for (v1, v2) in E:
            dist = 1 // our metric is number of hops, which is 1 for adjacent nodes
            if shortestPathLen[v2] >= shortestPathLen[v1] + dist:
                shortestPathLen[v2] = shortestPathLen[v1] + dist
                q.enqueue(v2)

diameter = 0
for v in V:
    shortestPathLen[v2] = +infinity, for all v2 in V
    BFS(v, shortestPathLen)
    // assuming network is connected, otherwise only reachable nodes
    longest = max(shortestPathLen)
    diameter = max(diameter, longest)
```

Since BFS runs in worst case complexity `O(|V| + |E|)` and we run it starting from every node,
we can get `O(|V|^2 + |V| |E|)`, which is quite a bit better.